utils/make_all.py

```diff
@@ -0,0 +1,166 @@
+    return all_list
+
+
+def write_inits(folder_path: str, all_dict: dict, skip_children: dict) -> dict:
+    """Write __init__.py files for all subfolders.
+
+    :return: folders with non-unique imports
+    """
+    folder_path = Path(folder_path)
+    non_unique = {}
+    for path in folder_path.glob('**/__init__.py'):
+        if path.parent == folder_path:
+            # skip top level __init__.py
+            continue
+
+        rel_path = path.parent.relative_to(folder_path).as_posix()
+
+        # get sub_dict for this folder
+        path_all_dict = all_dict
+        mod_path = path.parent.relative_to(folder_path).parts
+        try:
+            for part in mod_path:
+                path_all_dict = path_all_dict[part]
+        except KeyError:
+            # there is nothing to import
+            continue
+
+        path_all_dict = {key: val for key, val in path_all_dict.items() if key not in skip_children.get(rel_path, [])}
+        alls = gather_all(list(mod_path), path_all_dict, skip_children)
+
+        # check for non-unique imports
+        if len(alls + list(path_all_dict)) != len(set(alls + list(path_all_dict))):
+            non_unique[rel_path] = [k for k, v in Counter(alls + list(path_all_dict)).items() if v > 1]
+
+        auto_content = (['', '# AUTO-GENERATED'] + ['', '# yapf: disable', '# pylint: disable=wildcard-import', ''] +
+                        [f'from .{mod} import *' for mod in sorted(path_all_dict.keys())] + ['', '__all__ = ('] +
+                        [f'    {a!r},' for a in sorted(set(alls))] + [')', '', '# yapf: enable', ''])
+
+        start_content = []
+        end_content = []
+        in_docstring = in_end_content = False
+        in_start_content = True
+        for line in path.read_text(encoding='utf8').splitlines():
+            if not in_start_content and line.startswith('# END AUTO-GENERATED'):
+                in_end_content = True
+            if in_end_content:
+                end_content.append(line)
+                continue
+            # only use initial comments and docstring
+            if not (line.startswith('#') or line.startswith('"""') or in_docstring):
+                in_start_content = False
+                continue
+            if line.startswith('"""'):
+                if (not in_docstring) and not (line.endswith('"""') and not line.strip() == '"""'):
+                    in_docstring = True
+                else:
+                    in_docstring = False
+            if in_start_content and not line.startswith('# pylint'):
+                start_content.append(line)
+
+        new_content = start_content + auto_content + end_content
+
+        if not alls:
+            # there is nothing to import
+            continue
+
+        path.write_text('\n'.join(new_content).rstrip() + '\n', encoding='utf8')
+
+    return non_unique
+
+
+if __name__ == '__main__':
+    _folder = Path(__file__).parent.parent.joinpath('aiida')
+    _skip = {
```

Comment:
note also, here we specify children imports that should always be skipped

Reply:
please document in a comment why these imports are being skipped (individually, if there are different reasons, or once if they are all skipped for the same reason - e.g. import clashes that should be fixed).

Reply:
done 👍

Reply:
thanks! 

for the 'implementation' case - just to make sure I understand:
this is just a precaution, correct?
I.e. it would only take effect if we were to `from implementation import *`?

Reply:
yep thats correct: I tried to keep it as close to the existing imports as possible for now, but obviously we can iterate on exactly what is exposed where

Reply:
Ah, I didn't read the PR description carefully enough - I thought there might be some "priming" by the current imports, but what this PR does is to *impose* `from ... import *` for all submodules, correct?

I.e. there is no way to export a functionality only on aiida.x.y.z - it will always bubble up to aiida.x (unless it's in the skip list).
Is that correct?

Reply:
In a way, it means that we are using `__all__` (anywhere in the source tree) to define AiiiDA's public API. 

That is is ok, but I think we should document this and get input from some others to check they agree

Reply:
> Is that correct?

yep thats correct, and is basically what was already happening, just more manually 😬 (there was definitely places I think where people had just forgot to add certain imports)

> there is no way to export a functionality only on aiida.x.y.z - it will always bubble up to aiida.x (unless it's in the skip list)

currently yeh, although obviously now the basic code is in place, I could always make it fancier lol, for more fine-grained control